gstd/src/common/reservations.rs

```diff
@@ -0,0 +1,368 @@
+// This file is part of Gear.
+
+// Copyright (C) 2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Gear reservation manager implementation.
+//!
+//! It is capable of managing multiple gas reservations
+//! providing simple interface to user.
+
+use crate::{errors::Result, prelude::*};
+
+#[cfg(not(test))]
+use crate::exec;
+#[cfg(test)]
+use tests::exec_mock as exec;
+
+#[cfg(not(test))]
+use super::primitives::ReservationId;
+#[cfg(test)]
+use tests::ReservationIdMock as ReservationId;
+
+/// Stores additional data along with [`ReservationId`] to track its state.
+pub struct Reservation {
+    id: ReservationId,
+    amount: u64,
+    valid_until: u32,
+}
+
+impl Reservation {
+    /// Reserve the `amount` of gas for further usage.
+    ///
+    /// `duration` is the block count within which the reserve must be used.
+    ///
+    /// Refer to [`ReservationId`] for the more detailed description.
+    pub fn reserve(amount: u64, duration: u32) -> Result<Self> {
+        let block_height = exec::block_height();
+
+        Ok(Self {
+            id: ReservationId::reserve(amount, duration)?,
+            amount,
+            valid_until: duration.saturating_add(block_height),
+        })
+    }
+
+    /// Converts this [`Reservation`] into [`ReservationId`].
+    pub fn into_id(self) -> ReservationId {
```

Comment:
impl ReservationId from Reservation && provide getter by ref instead of just consume + derive some primitives for Reservation